launcher/src/code-workspace.ts

```diff
@@ -30,54 +44,103 @@ export class CodeWorkspace {
-        console.log(
-          `  > Workspace file ${workspaceFilePath()} is already exist`
-        );
-        return;
-      }
+      // get workspace file
+      const workspace = await this.readWorkspaceFile(alternativePath);
 
-      // take all the projects from flattened devworkspace file
-      const projects = await new FlattenedDevfile().getProjects();
-      const folders = projects.map((p) => {
-        return {
-          name: p.name,
-          path: `${env.PROJECTS_ROOT}/${p.name}`,
-        };
-      });
-
-      const workspace = {
-        folders,
-      };
-
-      // write .code-workspace file
-      const json = JSON.stringify(workspace, null, "\t");
-      await fs.writeFile(workspaceFilePath(), json);
+      const devfile = await new FlattenedDevfile().getDevfile();
 
-      console.log(`  > writing ${workspaceFilePath()}..`);
+      // synchronize projects
+      this.synchronizeProjects(workspace, devfile.projects);
+
+      // synchronize dependent projects
+      this.synchronizeProjects(workspace, devfile.dependentProjects);
+
+      // synchronize starter projects
+      this.synchronizeProjects(workspace, devfile.starterProjects);
+
+      // write workspace file
+      await this.writeWorkspaceFile(workspace, alternativePath);
     } catch (err) {
       console.error(
         `${err.message} Unable to generate che.code-workspace file`
       );
     }
   }
+
+  async isExist(workspaceFilePath: string | undefined): Promise<boolean> {
+    if (
+      workspaceFilePath &&
+      (await fs.pathExists(workspaceFilePath)) &&
+      (await fs.isFile(workspaceFilePath))
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  // should validate the content of read file
+  async readWorkspaceFile(alternativePath?: string): Promise<Workspace> {
+    let workspace;
+    if (await this.isExist(alternativePath)) {
+      workspace = JSON.parse(await fs.readFile(alternativePath!));
+    } else if (await this.isExist(workspaceFilePath())) {
+      workspace = JSON.parse(await fs.readFile(workspaceFilePath()));
+    } else {
+      workspace = {};
+    }
+
+    return workspace;
+  }
+
+  async writeWorkspaceFile(
+    workspace: Workspace,
+    alternativePath?: string
+  ): Promise<void> {
+    const json = JSON.stringify(workspace, null, "\t");
+
+    if (await this.isExist(alternativePath)) {
+      console.log(`  > writing ${alternativePath}..`);
+      await fs.writeFile(alternativePath!, json);
+    } else {
+      console.log(`  > writing ${workspaceFilePath()}..`);
+      await fs.writeFile(workspaceFilePath(), json);
+    }
+  }
+
+  async synchronizeProjects(
+    workspace: Workspace,
+    projects: Project[] | undefined
+  ) {
+    if (!projects) {
+      return;
+    }
+
+    if (!workspace.folders) {
+      workspace.folders = [];
+    }
+
+    const existInWorkspace = (projectName: string): boolean => {
+      for (var i = 0; i < workspace.folders.length; i++) {
```

Comment:
`for (let i=0...)

Reply:
btw
it can be simplified like:
`return workspace.folders.some(folder => folder.name === projectName);`